ports/esp8266/common-hal/pulseio/PulseIn.c

```diff
@@ -26,51 +26,161 @@
 
 #include <stdint.h>
 
+#include <user_interface.h>
+#include <eagle_soc.h>
+#include "esp_mphal.h"
+
 #include "mpconfigport.h"
 #include "py/gc.h"
 #include "py/runtime.h"
 #include "shared-bindings/microcontroller/__init__.h"
 #include "shared-bindings/pulseio/PulseIn.h"
+#include "common-hal/microcontroller/__init__.h"
+
+static void pulsein_set_interrupt(pulseio_pulsein_obj_t *self, bool rising, bool falling) {
+    ETS_GPIO_INTR_DISABLE();
+    // Set interrupt mode
+    GPIO_REG_WRITE(
+        GPIO_PIN_ADDR(self->pin->gpio_number),
+        (GPIO_REG_READ(GPIO_PIN_ADDR(self->pin->gpio_number) & ~GPIO_PIN_INT_TYPE_MASK)) |
+        GPIO_PIN_INT_TYPE_SET(
+            (rising ? GPIO_PIN_INTR_POSEDGE : 0) | (falling ? GPIO_PIN_INTR_NEGEDGE : 0)
+        )
+    );
+    // Clear interrupt status
+    GPIO_REG_WRITE(GPIO_STATUS_W1TC_ADDRESS, 1 << self->pin->gpio_number);
+    ETS_GPIO_INTR_ENABLE();
+}
+
+void pulseio_pulsein_interrupt_handler(void *data) {
+    pulseio_pulsein_obj_t *self = data;
+    uint32_t time_us = system_get_time();
+    if (self->first_edge) {
+        self->first_edge = false;
+        pulsein_set_interrupt(self, true, true);
+    } else {
+        uint16_t elapsed_us = (uint16_t)(time_us - self->last_us);
+        uint16_t i = (self->start + self->len) % self->maxlen;
+        self->buffer[i] = elapsed_us;
+        if (self->len < self->maxlen) {
+            self->len++;
+        } else {
+            self->start++;
+        }
+    }
+    self->last_us = time_us;
+}
 
 void common_hal_pulseio_pulsein_construct(pulseio_pulsein_obj_t* self,
         const mcu_pin_obj_t* pin, uint16_t maxlen, bool idle_state) {
-    mp_raise_NotImplementedError("");
+    if (pin->gpio_number == NO_GPIO || pin->gpio_function == SPECIAL_CASE) {
```

Comment:
Can we extend this to GPIO16? It is the only pin with `gpio_function == SPECIAL_CASE`.

Reply:
Unfortunately, GPIO16 has no IRQ capability so we can't practically use it for PulseIn ... it's a bit of a lame duck pin.

Reply:
Ahhh...that's right. The curse of being attached to the RTC register.